### PR TITLE
add isl_pw_multi_aff_range_factor_{domain,range}

### PR DESCRIPTION
**Pull request by skimo-openhub**

### CODE REVIEW

**Comment by tobig** [comment]
Sounds useful, indeed!

File: isl_aff.c [hunk 1]
```diff
@@ -4228,8 +4228,8 @@ __isl_give isl_set *isl_multi_aff_lex_gt_set(__isl_take isl_multi_aff *ma1,
 
 #undef PW
 #define PW isl_pw_multi_aff
-#undef EL
-#define EL isl_multi_aff
+#undef BASE
+#define BASE multi_aff
 #undef EL_IS_ZERO
```
Comment (ftynse) [inline isl_aff.c]: Given that `EL` is now defined automatically from `BASE`, maybe it's better to avoid using it in other macros?  E.g., rename `EL_IS_ZERO` into `IS_ZERO` ?

File: isl_pw_templ.c [hunk 2]
```diff
@@ -1562,6 +1562,71 @@ __isl_give PW *FN(PW,restore_space)(__isl_take PW *pw,
 	return NULL;
 }
 
+/* Return a copy of the function associated to
```
Comment (ftynse) [inline isl_pw_templ.c]: Is `pw` intended to necessarily be a function?  I understand it is the case now, but a couple of functions above or below this one avoid mentioning "function".

File: isl_pw_templ.c [hunk 3]
```diff
@@ -1562,6 +1562,71 @@ __isl_give PW *FN(PW,restore_space)(__isl_take PW *pw,
...
+ * the part of the domain at position "pos" in "pw".
+ */
+__isl_give EL *FN(FN(PW,get),BASE)(__isl_keep PW *pw, int pos)
+{
+	if (!pw)
+		return NULL;
+	return FN(EL,copy)(pw->p[pos].FIELD);
```
Comment (ftynse) [inline isl_pw_templ.c]: I would add a check on `pos` being non-negative and not overflowing.

File: isl_pw_templ.c [hunk 4]
```diff
@@ -1562,6 +1562,71 @@ __isl_give PW *FN(PW,restore_space)(__isl_take PW *pw,
...
+}
+
+/* Return the function associated to
+ * the part of the domain at position "pos" in "pw".
+ * This may be either a copy or the function itself
+ * if there is only one reference to "pw".
+ * This allows the function to be modified inplace
+ * if both the piecewise expression and this function
```
Comment (ftynse) [inline isl_pw_templ.c]: I don't understand what "this function" means in this context, and there does not seem to be a check that "this function has only a single reference" in the code.

File: isl_pw_templ.c [hunk 5]
```diff
@@ -1562,6 +1562,71 @@ __isl_give PW *FN(PW,restore_space)(__isl_take PW *pw,
...
+	if (!pw)
+		return NULL;
+	if (pw->ref != 1)
+		return FN(FN(PW,get),BASE)(pw, pos);
+	el = pw->p[pos].FIELD;
+	pw->p[pos].FIELD = NULL;
+	return el;
+}
+
+/* Return the function associated to
```
Comment (ftynse) [inline isl_pw_templ.c]: Isn't it rather "set the function associated to the part <...> and return the piecewise expression"?  Or even "set the expression associated tot he part <...> and return the updated piecewise expression" to avoid mentioning "function.

**Comment by skimo-openhub** [comment]
On Wed, Mar 28, 2018 at 02:26:04PM -0700, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -4228,8 +4228,8 @@ __isl_give isl_set *isl_multi_aff_lex_gt_set(__isl_take isl_multi_aff *ma1,
>  
>  #undef PW
>  #define PW isl_pw_multi_aff
> -#undef EL
> -#define EL isl_multi_aff
> +#undef BASE
> +#define BASE multi_aff
>  #undef EL_IS_ZERO
> 
> Given that `EL` is now defined automatically from `BASE`, maybe it's better to avoid using it in other macros?  E.g., rename `EL_IS_ZERO` into `IS_ZERO` ?

The original naming was probably not perfect, but I don't think
this change should affect the name of EL_IS_ZERO.
Btw, IS_ZERO already exists and it's not always the same as EL_IS_ZERO.
(Perhaps this should also be reconsidered, but not for this PR.)

skimo

**Comment by skimo-openhub** [comment]
On Wed, Mar 28, 2018 at 09:36:02PM +0000, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -1562,6 +1562,71 @@ __isl_give PW *FN(PW,restore_space)(__isl_take PW *pw,
>  	return NULL;
>  }
>  
> +/* Return a copy of the function associated to
> 
> Is `pw` intended to necessarily be a function?  I understand it is the case now, but a couple of functions above or below this one avoid mentioning "function".

I'll see if I can rephrase this together with your other two remarks.

> > @@ -1562,6 +1562,71 @@ __isl_give PW *FN(PW,restore_space)(__isl_take PW *pw,
>  	return NULL;
>  }
>  
> +/* Return a copy of the function associated to
> + * the part of the domain at position "pos" in "pw".
> + */
> +__isl_give EL *FN(FN(PW,get),BASE)(__isl_keep PW *pw, int pos)
> +{
> +	if (!pw)
> +		return NULL;
> +	return FN(EL,copy)(pw->p[pos].FIELD);
> 
> I would add a check on `pos` being non-negative and not overflowing.

This is only meant to be used internally (which is why I didn't add it),
but I guess it wouldn't hurt to add a check.

skimo
